File: leetcode/python/Spiral Matrix II/Spiral_Matrix_II.py
class Solution:
    # @return a list of lists of integer
    def generateMatrix(self, n):
        if n == 0:
            return []
        # res = [[0] * n] * n  Ç³¿½±´ £¬³ö´í
        res = [[0 for i in range(n)] for i in range(n)]
        left = 0
        top = 0
        right = n - 1
        bottom = n - 1
        direct = 0 # 0:go right 1:down 2:go left 3:up  
        val = 1
        while True :
            if direct == 0:
                for i in range(left,right + 1):
                    res[top][i] = val
                    val += 1
                top += 1
            if direct == 1:
                for i in range(top,bottom + 1):
                    res[i][right] = val
                    val += 1
                right -= 1
            if direct == 2:
                for i in range(right,left -1 ,-1):
                    res[bottom][i] = val
                    val += 1
                bottom -= 1
            if direct == 3:
                for i in range(bottom,top - 1,-1):
                    res[i][left] = val
                    val += 1
                left += 1
            direct = (direct + 1) % 4
            if top > bottom or left > right:
                return res

solu = Solution()
solu.generateMatrix(3)

File: leetcode/cpp/SpiralMatrixII.cpp
//
// Created by zz on 2017/10/8.
// Time : O(n^2)
// Space : O(1)

#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> ans(n, vector<int>(n));
        int num = 0;
        for (int left = 0, right = n - 1, top = 0, bottom = n - 1;
             left <= right && top <= bottom;
             left++, right--, top++, bottom--) {

            // 从左到右
            for (int i = left; i <= right; i++) {
                ans[top][i] = ++num;
            }

            // 从上到下
            for (int i = top + 1; i <= bottom; i++) {
                ans[i][right] = ++num;
            }

            // 从右到左
            for (int i = right - 1; top < bottom && i >= left; i--) {
                ans[bottom][i] = ++num;
            }

            // 从下到上
            for (int i = bottom - 1; left < right && i >= top + 1; i--) {
                ans[i][left] = ++num;
            }
        }
        return ans;


    }
};